### PR TITLE
More android fixes and additions

### PR DESCRIPTION
**Pull request by hhrowal**
Fixed the signature for a couple callback methods. Fixed local reference table overflow. Added missing methods to AbstractPart.

### CODE REVIEW

File: src/java/com/libmailcore/HTMLRendererTemplateCallback.java [hunk 1]
```diff
@@ -12,9 +12,9 @@
     /** Returns true if an attachment should (such as an image) be shown. */
     boolean shouldShowPart(AbstractPart part);
     /** Returns the values used in the header template for a given template. */
-    Map<String, String> templateValuesForHeader(MessageHeader header);
+    Map<String, Object> templateValuesForHeader(MessageHeader header);
```
Comment (dinhviethoa) [inline src/java/com/libmailcore/HTMLRendererTemplateCallback.java]: Could you tell me why you changed it to Object? Values are always String.

File: src/java/TypesUtils.cpp [hunk 2]
```diff
@@ -342,6 +342,7 @@ static bool isJavaMap(JNIEnv * env, jobject obj)
 
 static Object * hashmapJavaToObjectConverter(JNIEnv * env, jobject obj)
 {
+	env->PushLocalFrame(20);
```
Comment (dinhviethoa) [inline src/java/TypesUtils.cpp]: Could you let me know why it's required?

**Comment by dinhviethoa** [comment]
Some comments, everything else looks good!

File: src/java/TypesUtils.cpp (hunk 2 above)
Comment (hhrowal) [inline src/java/TypesUtils.cpp]: The local object references are usually cleaned up when control is returned to the JVM. However, during the htmlRendering operation, numerous calls are made through the JNI to gather the data necessary for rendering. Control is not returned to the JVM until all the callback methods have been executed. So much data is returned from the callbacks that he local object reference table fills up (512 entries) and causes app termination. This just releases the object references as soon as they are no longer required.

File: src/java/com/libmailcore/HTMLRendererTemplateCallback.java (hunk 1 above)
Comment (hhrowal) [inline src/java/com/libmailcore/HTMLRendererTemplateCallback.java]: The default callback in src/core/renderer/MCHTMLRendererCallback.cpp returns empty adds empty HashMaps to the result.

result->setObjectForKey(MCSTR("NOFROM"), mailcore::HashMap::hashMap());

In order to mimic that behavior, it was necessary to return a Map<String, Object>.

Comment (hhrowal) [inline src/java/com/libmailcore/HTMLRendererTemplateCallback.java]: My comment was "edited". It is necessary to return `Map<String, Object> `

File: src/java/TypesUtils.cpp (hunk 2 above)
Comment (dinhviethoa) [inline src/java/TypesUtils.cpp]: What does 20 or 10 means? is it going to be the limit?
If that's just a hint and not the limit, could you use the same value everywhere (and maybe use a constant)?

File: src/java/TypesUtils.cpp [hunk 3]
```diff
@@ -351,6 +352,7 @@ static Object * hashmapJavaToObjectConverter(JNIEnv * env, jobject obj)
     jobjectArray array = (jobjectArray) env->CallObjectMethod(entrySet, method);
     int count = (int) env->GetArrayLength(array);
     for(int i = 0 ; i < count ; i ++) {
+		env->PushLocalFrame(10);
```
Comment (dinhviethoa) [inline src/java/TypesUtils.cpp]: Could you fix indentation here?

File: src/java/TypesUtils.cpp (hunk 2 above)
Comment (dinhviethoa) [inline src/java/TypesUtils.cpp]: There's also an indentation issue.

Comment (hhrowal) [inline src/java/TypesUtils.cpp]: I see now there are other places in TypesUtils.cpp that already call PushLocalFrame with a value of 32. I will create a constant and make sure all calls use that constant

File: src/java/TypesUtils.cpp (hunk 3 above)
Comment (hhrowal) [inline src/java/TypesUtils.cpp]: Created LOCAL_FRAME_CAPACITY constant and fixed indentation.

**Comment by dinhviethoa** [comment]
Thanks for helping make the Android support a real thing!

**Pull request merged**